File: src/domain/services/applicationService.go
package services

import (
	"Theatrum/constants"
	"Theatrum/domain/models"
	"Theatrum/domain/repositories"
	"Theatrum/domain/utils"
	"fmt"
	"log"
	"path"
	"strings"
)

// ApplicationService handles application-wide operations and configuration
type ApplicationService struct {
	application    *models.Application
	server         *models.Server
	channels       *map[string]models.Stream
	storage        repositories.StoragePort
	templateService *PathTemplateService
}

// NewApplicationService creates a new instance of ApplicationService
func NewApplicationService(application *models.Application, server *models.Server, channels *map[string]models.Stream, storage repositories.StoragePort, templateService *PathTemplateService) *ApplicationService {
	applicationService := &ApplicationService{
		application:    application,
		server:        server,
		channels:      channels,
		storage:       storage,
		templateService: templateService,
	}
	return applicationService
}

// GetApplication returns the application configuration
func (s *ApplicationService) GetApplication() *models.Application {
	return s.application
}

// GetServer returns the server configuration
func (s *ApplicationService) GetServer() *models.Server {
	return s.server
}

// GetChannels returns all channels
func (s *ApplicationService) GetChannels() *map[string]models.Stream {
	return s.channels
}

// GetChannel retrieves a specific channel by ID
func (s *ApplicationService) GetChannel(channelId string) (models.Stream, error) {
	channels := s.GetChannels()
	channel, exists := (*channels)[channelId]
	if !exists {
		return models.Stream{}, fmt.Errorf("channel not found: %s", channelId)
	}
	return channel, nil
}

// GetAllStreamsPlaylist generates an M3U8 playlist containing all available streams
func (s *ApplicationService) BuildAllStreamsPlaylist() (string, error) {
	if !s.application.AllStreamsPlaylist.Enabled {
		return "", fmt.Errorf("all streams playlist is not enabled")
	}

	// Start building the M3U8 playlist
	var playlist strings.Builder

	// Write M3U8 header
	playlist.WriteString("#EXTM3U\n")
	playlist.WriteString("#EXT-X-VERSION:3\n")
	playlist.WriteString("#EXT-X-STREAM-INF:BANDWIDTH=0\n")

	// Search for the MasterPlaylist file for each streams
	for index, stream := range *s.channels {
		patternMasterPlaylist := path.Join(constants.VideoDir, stream.GetMasterPlaylistTemplatePath())
		masterFiles, vars, err := s.storage.SearchFiles(patternMasterPlaylist, constants.ValidMasterPlaylistExtensions)
		if err != nil {
			log.Printf("Error searching for videos in %s: %v", stream.Path, err)
			continue
		}

		// Search in the filename that there is a constant.MasterPlaylist
		for i, _ := range masterFiles {
			// Get the public path of the master playlist
			channelPath, err := s.templateService.ReplacePlaceholders(index, vars[i])
			if err != nil {
				log.Printf("Error replacing placeholders: %v", err)
				continue
			}

			masterFilePublicPath := utils.JoinURL(s.application.PublicPath, channelPath, constants.MasterPlaylist)

			// Add the master playlist to the playlist
			log.Println("Adding master playlist:", masterFilePublicPath)
			playlist.WriteString(fmt.Sprintf("%s\n", masterFilePublicPath))
		}
	}

	return playlist.String(), nil
}

// Cleanup performs any necessary cleanup operations
func (s *ApplicationService) Cleanup() {
	// Add any cleanup operations here
	// For example, closing any open connections, files, etc.
}

File: src/main/java/com/fjourdren/theatrum/domain/service/ApplicationService.java
package com.fjourdren.theatrum.domain.service;

import com.fjourdren.theatrum.application.port.in.ResolveChannelUseCase;
import com.fjourdren.theatrum.application.port.out.StoragePort;
import com.fjourdren.theatrum.domain.constant.VideoConstants;
import com.fjourdren.theatrum.domain.model.AppPaths;
import com.fjourdren.theatrum.domain.model.Application;
import com.fjourdren.theatrum.domain.model.FileMatch;
import com.fjourdren.theatrum.domain.model.LoadedConfiguration;
import com.fjourdren.theatrum.domain.model.Server;
import com.fjourdren.theatrum.domain.model.Stream;
import com.fjourdren.theatrum.domain.util.UrlUtils;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;

/**
 * Application-wide operations and configuration.
 *
 * <p>{@link LoadedConfiguration} is parsed before the context exists and registered as a
 * pre-existing singleton by {@code TheatrumApplication}, so it injects like any other bean.
 */
@Component
@Slf4j
@RequiredArgsConstructor
public class ApplicationService implements ResolveChannelUseCase {

    private final LoadedConfiguration configuration;
    private final StoragePort storage;
    private final PathTemplateService templateService;
    private final AppPaths appPaths;

    @Override
    public Application getApplication() {
        return configuration.application();
    }

    @Override
    public Server getServer() {
        return configuration.server();
    }

    @Override
    public Map<String, Stream> getChannels() {
        return configuration.channels();
    }

    /** @throws IllegalArgumentException when no channel is configured under {@code channelId} */
    @Override
    public Stream getChannel(String channelId) {
        Stream channel = getChannels().get(channelId);
        if (channel == null) {
            throw new IllegalArgumentException("channel not found: " + channelId);
        }
        return channel;
    }

    /**
     * Builds an M3U8 playlist listing the master playlist of every available stream.
     *
     * @throws IllegalStateException when the all-streams playlist is disabled
     */
    @Override
    public String buildAllStreamsPlaylist() {
        if (!getApplication().allStreamsPlaylist().enabled()) {
            throw new IllegalStateException("all streams playlist is not enabled");
        }

        var playlist = new StringBuilder("""
                #EXTM3U
                #EXT-X-VERSION:3
                #EXT-X-STREAM-INF:BANDWIDTH=0
                """);

        // Search for the master playlist file of each stream.
        for (var entry : getChannels().entrySet()) {
            Stream stream = entry.getValue();
            String pattern = appPaths.videoDir().resolve(stream.getMasterPlaylistTemplatePath()).toString();

            List<FileMatch> matches;
            try {
                matches = storage.searchFiles(pattern, VideoConstants.VALID_MASTER_PLAYLIST_EXTENSIONS);
            } catch (Exception e) {
                log.error("Error searching for videos in {}: {}", stream.path(), e.getMessage());
                continue;
            }

            for (FileMatch match : matches) {
                String channelPath;
                try {
                    channelPath = templateService.replacePlaceholders(entry.getKey(), match.vars());
                } catch (RuntimeException e) {
                    log.error("Error replacing placeholders: {}", e.getMessage());
                    continue;
                }

                String masterFilePublicPath = UrlUtils.joinUrl(
                        getApplication().publicPath(), channelPath, VideoConstants.MASTER_PLAYLIST);

                log.info("Adding master playlist: {}", masterFilePublicPath);
                playlist.append(masterFilePublicPath).append('\n');
            }
        }

        return playlist.toString();
    }
}
